orchestrator/careplanservice/handle_createtask.go

```diff
@@ -30,17 +34,74 @@ func (s *Service) handleCreateTask(httpRequest *http.Request, tx *coolfhir.Trans
 	default:
 		return nil, errors.New(fmt.Sprintf("cannot create Task with status %s, must be %s or %s", task.Status, fhir.TaskStatusRequested.String(), fhir.TaskStatusReady.String()))
 	}
-	// Resolve the CarePlan
-	carePlanRef, err := basedOn(task)
+
+	var carePlan fhir.CarePlan
+	var carePlanRef *string
+	err := coolfhir.ValidateTaskRequiredFields(task)
 	if err != nil {
-		//FIXME: This logic changed, create a new CarePlan when Task.basedOn is not set
-		return nil, fmt.Errorf("invalid Task.basedOn: %w", err)
+		return nil, err
 	}
-	// TODO: Manage time-outs properly
-	var carePlan fhir.CarePlan
-	if err := s.fhirClient.Read(*carePlanRef, &carePlan); err != nil {
-		return nil, fmt.Errorf("failed to read CarePlan: %w", err)
+	// Resolve the CarePlan
+	if len(task.BasedOn) == 0 {
+		// The CarePlan does not exist, a CarePlan and CareTeam will be created and the requester will be added as a member
+		cp := fhir.CarePlan{}
+
+		carePlanBytes, _ := json.Marshal(cp)
+		newHttpRequest := httptest.NewRequest("POST", "/CarePlan", bytes.NewReader(carePlanBytes))
+
+		// Create a new transaction, we need these resources to be created before the flow can continue
+		newTx := coolfhir.Transaction()
+
+		res, err := s.handleCreateCarePlan(newHttpRequest, newTx)
+		if err != nil {
+			return nil, err
+		}
+		if res == nil {
+			return nil, errors.New("invalid response")
+		}
+
+		bundle, err := s.commitTransaction(newHttpRequest, newTx, []FHIRHandlerResult{res})
+		if err != nil {
+			return nil, err
+		}
+
+		err = coolfhir.ResourceInBundle(bundle, coolfhir.EntryIsOfType("CarePlan"), &carePlan)
+		if err != nil {
+			return nil, err
+		}
+
+		task.BasedOn = []fhir.Reference{
+			{
+				Type:      to.Ptr("CarePlan"),
+				Reference: to.Ptr("CarePlan/" + *carePlan.Id),
+			},
+		}
+	} else {
+		// Adding a task to an existing CarePlan
+		carePlanRef, err = basedOn(task)
```

Comment:
We should be starting to write unit tests, as we're working towards production-ready code (and relying on slow integration tests will ultimately slow down development). Should be easy, since the primary output of these handlers is now the Bundles entries they add to the transaction (`tx`). See `handle_createcareplan_test.go` for an example (`TestService_handleCreateCarePlan`)

Reply:
Yes that's a fair point! Unit tests were still on the way but I wanted to prioritise on getting the flow working with the integ tests to also start with the review